sdk/cosmosdb/cosmos/src/queryExecutionContext/pipelinedQueryExecutionContext.ts

```diff
@@ -195,30 +273,115 @@ export class PipelinedQueryExecutionContext implements ExecutionContext {
-            this.fetchBuffer = this.fetchBuffer.slice(this.pageSize);
-            return { result: temp, headers: this.fetchMoreRespHeaders };
-          } else {
-            const temp = this.fetchBuffer;
-            this.fetchBuffer = [];
-            return { result: temp, headers: this.fetchMoreRespHeaders };
-          }
-        }
-        // Recursively fetch more results to ensure the pageSize number of results are returned
-        // to maintain compatibility with the previous implementation
-        return this._fetchMoreImplementation(diagnosticNode);
+      // Return collected items up to pageSize
+      if (this.fetchBuffer.length > 0) {
+        const temp = this.fetchBuffer.slice(0, this.pageSize);
+        this.fetchBuffer = this.fetchBuffer.slice(this.pageSize);
+        return { result: temp, headers: this.fetchMoreRespHeaders };
+      } else {
+        return { result: undefined, headers: this.fetchMoreRespHeaders };
       }
     } catch (err: any) {
       mergeHeaders(this.fetchMoreRespHeaders, err.headers);
       err.headers = this.fetchMoreRespHeaders;
-      if (err) {
-        throw err;
-      }
+      throw err;
+    }
+  }
+
+  private async _enableQueryControlFetchMoreImplementation(
+    diagnosticNode: DiagnosticNodeInternal,
+  ): Promise<Response<any>> {
+    // Use continuation token logic if manager was created (supported queries)
+    // Otherwise use simplified buffer-only logic (unsupported queries)
+    if (this.continuationTokenManager) {
+      return this._handleQueryFetch(diagnosticNode);
+    } else {
+      return this._handleSimpleBufferFetch(diagnosticNode);
+    }
+  }
+
+  private async _handleSimpleBufferFetch(
+    diagnosticNode: DiagnosticNodeInternal,
+  ): Promise<Response<any>> {
+    // Return buffered data if available
+    if (this.fetchBuffer.length > 0) {
+      const temp = this.fetchBuffer.slice(0, this.pageSize);
+      this.fetchBuffer = this.fetchBuffer.slice(this.pageSize);
+      return { result: temp, headers: this.fetchMoreRespHeaders };
+    }
+
+    // Fetch new data from endpoint
+    const response = await this.endpoint.fetchMore(diagnosticNode);
+    mergeHeaders(this.fetchMoreRespHeaders, response.headers);
+
+    if (!response?.result?.buffer?.length) {
+      return this.createEmptyResult(response?.headers);
     }
+
+    // Buffer new data and return up to pageSize
+    this.fetchBuffer = response.result.buffer;
+    const temp = this.fetchBuffer.slice(0, this.pageSize);
+    this.fetchBuffer = this.fetchBuffer.slice(this.pageSize);
+
+    return { result: temp, headers: this.fetchMoreRespHeaders };
+  }
+
+  private async _handleQueryFetch(diagnosticNode: DiagnosticNodeInternal): Promise<Response<any>> {
+    if (this.fetchBuffer.length > 0) {
+      const { endIndex, continuationToken } = this.continuationTokenManager.paginateResults(
+        this.pageSize,
+        this.fetchBuffer.length === 0,
```

Comment:
Is this expected? 
this is only executed when length > 0

Reply:
yeah, but it will always be false. simplifying the condition